docs/engineering/Testing/ui-tests.md

```diff
@@ -0,0 +1,550 @@
+  propA: "third value",
+  propD: value,
+};
+```
+
+The goal is to have enough stories to display each state a component could be in and each option it could have. If you are writing tests for an existing component, it should already have stories written, but you may find that you have to tweak them a bit, or add additional props, to cover all the cases you need to test.
+
+:::tip Typescript Tip
+If you ever want to see exactly what all the props a given component should or could have are, typescript has your back. Look in the src file for the component and find the `interface` and/or `type` declarations where the types are defined for the component's props. Here's an example from the BoardMenuItem component of Boba's Side Menu:
+
+```typescript
+export interface BoardMenuItemProps {
+  avatar: string;
+  color: string;
+  updates?: number | boolean;
+  slug: string;
+  link: LinkWithAction;
+  muted?: boolean;
+  outdated?: boolean;
+  current?: boolean;
+  loading?: false;
+}
+```
+
+:::
+
+### Actions
+
+If a component has props that are links or functions, we can use [actions](https://storybook.js.org/docs/react/essentials/actions) to model what happens when they are triggered.
+
+For example, for a link, your `Story.args` might include:
+
+```typescript
+link: { href: "#slug", onClick: action("#slug") },
+```
+
+and a component with a Submit button on a text input might have:
+
+```typescript
+onSubmit: (text) => action("submit")(text),
+```
+
+When you type in this text input in Storybook and click the Submit button, the callback will be logged in Storybook's Action tab with _submit: "the text you typed"_. We can also leverage these actions in our tests to simulate what happens when a user interacts with the component.
+
+## Jest Tests with React Testing Library
+
+[Jest](https://jestjs.io/) tests let us compare our expected outcome for a piece of code to the actual result.
+
+[Testing Library](https://testing-library.com/docs/dom-testing-library/intro) (and specifically its extension [React Testing Library](https://testing-library.com/docs/react-testing-library/intro)) renders our components to the DOM and then lets us find specific DOM nodes with queries that mimic how real users would interact with them.
+
+Together they let us compare UI elements as they are rendered to what we expect to be there and how we expect them to work.
+
+### Setting Up the Test File
+
+The first step in writing tests is to set up your test file. You can use the following as a template:
+
+```typescript title="tests/11-ComponentGroup/Component.test.tsx"
+import "@testing-library/jest-dom/extend-expect";
+
+// Import the stories you're going to test against
+import * as stories from "stories/11-ComponentGroup/01-Component.stories";
+
+// Import the prop types for the component
+import { ComponentProps } from "../../src/path/Component";
+
+import { render, screen, waitFor, within } from "@testing-library/react";
+
+import React from "react";
```

Comment:
Don't introduce imports before they're necessary. If we want to have a full testing template, this should be in the UI tests codebase, not in the documentation (though we can import it in the documentation from there). As a rule of thumb, the further away from the code something like this is, the less likely it is to be updated when changes happen.

I don't think a test template is necessary as of now.